### PR TITLE
Adding control of leaf water potential on transpiration through inner leaf humidity (Hydro only)

### PR DESCRIPTION
**Pull request by JunyanDing**
### Description:

 
Issue #726

### Collaborators:

Ryan Knox, Chonggang Xu, Charlie Koven

### Expectation of Answer Changes:
1. incorporated the lwp on transpiration (see attached file for detail) 
[StomaModification.pdf](https://github.com/NGEET/fates/files/6281068/StomaModification.pdf)
2.  provide an option to use agross in BB stoma conductance model controlled by in code parameter  use_agross

### Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the in-code documentation .AND. (the [technical note](https://github.com/NGEET/fates-docs) .OR. the wiki) accordingly.
- [ x] I have read the [**CONTRIBUTING**](https://github.com/NGEET/fates/blob/master/CONTRIBUTING.md) document.
- [ ] FATES PASS/FAIL regression tests were run
- [ ] If answers were expected to change, evaluation was performed and provided

### Test Results:

CTSM (or) E3SM (specify which) test hash-tag:

CTSM (or) E3SM (specify which) baseline hash-tag:

FATES baseline hash-tag:

Test Output:

### CODE REVIEW

File: biogeophys/FatesPlantRespPhotosynthMod.F90 [hunk 1]
```diff
@@ -1214,17 +1268,75 @@ subroutine LeafLayerPhotosynthesis(f_sun_lsl,         &  ! in
...
+             ! now adjust veg_esat by LWP_star
+             veg_esat_adj = veg_esat*LWP_star
+
+             ! note: q is the specific humidity
+             qs = 0.622 * ceair / (can_press - 0.378 * ceair)
+             qsat = 0.622 * veg_esat / (can_press - 0.378 * veg_esat)
+             ! qsat_adj = 0.622 * veg_esat_adj / (can_press - 0.378 * veg_esat_adj)
+             qsat_adj = qsat*LWP_star
+             
+             !  write (fates_log(),*) 'qsat:', qsat, 'RHleaf:', qsat_adj/qsat, 'RHs:', qs/qsat
```
Comment (rosiealice) [inline biogeophys/FatesPlantRespPhotosynthMod.F90:1312]:
```suggestion
             !  write (fates_log(),*) 'qsat:', qsat, 'RHleaf:', qsat_adj/qsat, 'RHs:', qs/qsat
```

File: biogeophys/FatesPlantRespPhotosynthMod.F90 [hunk 2]
```diff
@@ -1214,17 +1268,75 @@ subroutine LeafLayerPhotosynthesis(f_sun_lsl,         &  ! in
...
+             ! Vh20:  molar volume of water (18 × 10−6 m3 mol−1) 
+             ! R is the universal gas constant and T the interfacial temperature 
+             ! 
+             ! th_sat = EDPftvarcon_inst%hydr_thetas_node(ft,1)
+             ! th_rs = EDPftvarcon_inst%hydr_resid_node(ft,1)
+             
+             k_lwp = EDPftvarcon_inst%hydr_k_lwp(ft)
+             if (psi<0) then
+               LWP_star = exp(((k_lwp) * psi)*18.0_r8/(8.314_r8*veg_tempk))
+               ! k_lwp**btran
```
Comment (rosiealice) [inline biogeophys/FatesPlantRespPhotosynthMod.F90:1299]:
```suggestion
```

File: biogeophys/FatesPlantRespPhotosynthMod.F90 [hunk 3]
```diff
@@ -1214,17 +1268,75 @@ subroutine LeafLayerPhotosynthesis(f_sun_lsl,         &  ! in
...
+             else 
+               LWP_star = 1
+             end if
+             ! now adjust veg_esat by LWP_star
+             veg_esat_adj = veg_esat*LWP_star
+
+             ! note: q is the specific humidity
+             qs = 0.622 * ceair / (can_press - 0.378 * ceair)
+             qsat = 0.622 * veg_esat / (can_press - 0.378 * veg_esat)
+             ! qsat_adj = 0.622 * veg_esat_adj / (can_press - 0.378 * veg_esat_adj)
```
Comment (rosiealice) [inline biogeophys/FatesPlantRespPhotosynthMod.F90:1309]:
```suggestion
```

File: biogeophys/FatesPlantRespPhotosynthMod.F90 [hunk 4]
```diff
@@ -1214,17 +1268,75 @@ subroutine LeafLayerPhotosynthesis(f_sun_lsl,         &  ! in
...
+             if (psi<0) then
+               LWP_star = exp(((k_lwp) * psi)*18.0_r8/(8.314_r8*veg_tempk))
+               ! k_lwp**btran
+             else 
+               LWP_star = 1
+             end if
+             ! now adjust veg_esat by LWP_star
+             veg_esat_adj = veg_esat*LWP_star
+
+             ! note: q is the specific humidity
```
Comment (rosiealice) [inline biogeophys/FatesPlantRespPhotosynthMod.F90:1306]: Can you add a comment to say what is happening here?

File: biogeophys/FatesPlantRespPhotosynthMod.F90 [hunk 5]
```diff
@@ -1214,17 +1268,75 @@ subroutine LeafLayerPhotosynthesis(f_sun_lsl,         &  ! in
...
+             qsat = 0.622 * veg_esat / (can_press - 0.378 * veg_esat)
+             ! qsat_adj = 0.622 * veg_esat_adj / (can_press - 0.378 * veg_esat_adj)
+             qsat_adj = qsat*LWP_star
+             
+             !  write (fates_log(),*) 'qsat:', qsat, 'RHleaf:', qsat_adj/qsat, 'RHs:', qs/qsat
+             if (k_lwp == 0 ) then 
+               rstoma_out = 1._r8/gstoma
+             else 
+               if ((qsat_adj - qs) <= 0) then
+               !  write (fates_log(),*) 'LWP_star:', LWP_star
```
Comment (rosiealice) [inline biogeophys/FatesPlantRespPhotosynthMod.F90:1317]:
```suggestion
```

File: biogeophys/FatesPlantRespPhotosynthMod.F90 [hunk 6]
```diff
@@ -1214,17 +1268,75 @@ subroutine LeafLayerPhotosynthesis(f_sun_lsl,         &  ! in
...
+               !  write (fates_log(),*) 'qleaf:', qsat * LWP_star, 'qs',qs
```
Comment (rosiealice) [inline biogeophys/FatesPlantRespPhotosynthMod.F90:1318]:
```suggestion           
```

File: biogeophys/FatesPlantRespPhotosynthMod.F90 [hunk 7]
```diff
@@ -1214,17 +1268,75 @@ subroutine LeafLayerPhotosynthesis(f_sun_lsl,         &  ! in
...
+               !  write (fates_log(),*) 'LWP_star:', LWP_star
+               !  write (fates_log(),*) 'qleaf:', qsat * LWP_star, 'qs',qs               
+                 ! if inner leaf vapor pressure is less then or equal to that at leaf surface
+                 ! then set stomata resistance to be very large to stop the transpiration or back flow of vapor
+                  rstoma_out = rsmax0*100
+               else
+                  rstoma_out = (qsat-qs)*( 1/gstoma + rb)/(qsat_adj - qs)-rb
+                
+               end if
+               if (rstoma_out <=0 ) then
```
Comment (rosiealice) [inline biogeophys/FatesPlantRespPhotosynthMod.F90:1326]: Do you want to keep this check?

File: biogeophys/FatesPlantRespPhotosynthMod.F90 [hunk 8]
```diff
@@ -1214,17 +1268,75 @@ subroutine LeafLayerPhotosynthesis(f_sun_lsl,         &  ! in
...
+             ! LWP_star = wi/w0 = exp(psi_meso*Vh20/(R*T)) , w is the water vapor concentration 
+             ! Vh20:  molar volume of water (18 × 10−6 m3 mol−1) 
+             ! R is the universal gas constant and T the interfacial temperature 
+             ! 
+             ! th_sat = EDPftvarcon_inst%hydr_thetas_node(ft,1)
+             ! th_rs = EDPftvarcon_inst%hydr_resid_node(ft,1)
+             
+             k_lwp = EDPftvarcon_inst%hydr_k_lwp(ft)
+             if (psi<0) then
+               LWP_star = exp(((k_lwp) * psi)*18.0_r8/(8.314_r8*veg_tempk))
```
Comment (rosiealice) [inline biogeophys/FatesPlantRespPhotosynthMod.F90:1298]: Add a note to say that this will be 1 if the parameter value is zero?

File: biogeophys/FatesPlantRespPhotosynthMod.F90 [hunk 9]
```diff
@@ -1109,16 +1131,26 @@ subroutine LeafLayerPhotosynthesis(f_sun_lsl,         &  ! in
...
 
-                 ! Net carbon assimilation. Exit iteration if an < 0
-                 anet = agross  - lmr
-                 if (anet < 0._r8) then
-                    loop_continue = .false.
-                 end if
+
+                !!! Calculate anet, only exit iteration with negative anet when using anet in calculating gs
+                ! this is version B  
+                 anet = agross - lmr
```
Comment (rosiealice) [inline biogeophys/FatesPlantRespPhotosynthMod.F90:1137]: I guess I am a bit confused about why this section on agros is in this PR? Is it allied to the modification of gs by LWP, or orthogonal to it? If it's a seperate issue, should it be in a separate PR?

Comment (JunyanDing) [reply biogeophys/FatesPlantRespPhotosynthMod.F90:1137]: Sort of. If using lwp to constrain the transpiration, it is more appropriate to use agross instead of anet

File: biogeophys/FatesPlantRespPhotosynthMod.F90 (hunk 4 above)
Comment (JunyanDing) [reply biogeophys/FatesPlantRespPhotosynthMod.F90:1306]: I should delete veg_esat_adj and corresponding codes as this variable is no longer used.

File: biogeophys/FatesPlantRespPhotosynthMod.F90 (hunk 7 above)
Comment (JunyanDing) [reply biogeophys/FatesPlantRespPhotosynthMod.F90:1326]: logically, this rstoma_out would never be negative given the formula above, but I think it is be good to keep this check just in case.

File: biogeophys/FatesPlantRespPhotosynthMod.F90 [hunk 10]
```diff
@@ -1214,17 +1267,70 @@ subroutine LeafLayerPhotosynthesis(f_sun_lsl,         &  ! in
...
+             if (k_lwp == 0 ) then 
+               rstoma_out = 1._r8/gstoma
+             else 
+               if ((qsat_adj - qs) <= 0) then
+          
+                 ! if inner leaf vapor pressure is less then or equal to that at leaf surface
+                 ! then set stomata resistance to be very large to stop the transpiration or back flow of vapor
+                  rstoma_out = rsmax0*100
+               else
+                  rstoma_out = (qsat-qs)*( 1/gstoma + rb)/(qsat_adj - qs)-rb
```
Comment (xuchongang) [inline biogeophys/FatesPlantRespPhotosynthMod.F90:1317]: I personally do not like the add-on adjustment of leaf humidity on gs. I would prefer to updated the calculation of vpd (vapor pressure deficit from internal leave to air) or relative humidity (humidity in the air/humidity inside leave) for the two type of models. However, I think it is also justifiable as the above two models are fitted to air VPD assuming full saturation inside leaves. Not clear preference.....

File: biogeophys/FatesPlantRespPhotosynthMod.F90 [hunk 11]
```diff
@@ -216,7 +216,7 @@ subroutine FatesPlantRespPhotosynthDrive (nsites, sites,bc_in,bc_out,dtime)
                                    ! above the leaf layer of interest
     real(r8) :: lai_current        ! the LAI in the current leaf layer
     real(r8) :: cumulative_lai     ! the cumulative LAI, top down, to the leaf layer of interest
-
+    real(r8) :: psi_leaf           ! xylem water potential of leaf
```
[the remark this answers is not in the document]
Comment (JunyanDing) [reply biogeophys/FatesPlantRespPhotosynthMod.F90:219]: yes, will correct it

File: biogeophys/FatesPlantRespPhotosynthMod.F90 [hunk 12]
```diff
@@ -1214,17 +1267,70 @@ subroutine LeafLayerPhotosynthesis(f_sun_lsl,         &  ! in
...
+             else 
+               LWP_star = 1
+             end if
+             ! now adjust inner leaf humidity by LWP_star
+             ! note: q is the specific humidity
+             qs = 0.622 * ceair / (can_press - 0.378 * ceair)
+             qsat = 0.622 * veg_esat / (can_press - 0.378 * veg_esat)
+             qsat_adj = qsat*LWP_star
+             
+             if (k_lwp == 0 ) then
```
[the remark this answers is not in the document]
Comment (JunyanDing) [reply biogeophys/FatesPlantRespPhotosynthMod.F90:1289]: I was treating k_lwp as an positive integer in my mind when coding. Will adjust this.

File: biogeophys/FatesPlantRespPhotosynthMod.F90 [hunk 13]
```diff
@@ -530,7 +530,9 @@ subroutine FatesPlantRespPhotosynthDrive (nsites, sites,bc_in,bc_out,dtime)
                                                         bc_in(s)%forc_pbot,                 &  ! in
                                                         bc_in(s)%cair_pa(ifp),              &  ! in
                                                         bc_in(s)%oair_pa(ifp),              &  ! in
+                                                        bc_in(s)%rb_pa(ifp),                &  ! in                                                        
                                                         btran_eff,                          &  ! in
+                                                        psi_leaf,                           &  ! in
```
Comment (rgknox) [inline biogeophys/FatesPlantRespPhotosynthMod.F90:535]: I believe we call this routine for both hydro and non-hydro versions of the code.  One solution is to initialize psi_leaf with a flag value for non-hydro cases, so we can option to turn off its use when we enter LeafLayerPhotosynthesis()

File: biogeophys/FatesPlantRespPhotosynthMod.F90 [hunk 14]
```diff
@@ -1131,23 +1159,30 @@ subroutine LeafLayerPhotosynthesis(f_sun_lsl,         &  ! in
...
                           (2.0*stomatal_intercept_btran + term * &
                           (1.0 - medlyn_slope(ft)* medlyn_slope(ft) / vpd)) * term
-           
+
                   call quadratic_f (aquad, bquad, cquad, r1, r2)
                   gs_mol = max(r1,r2)
 
-                else if ( stomatal_model == 1 ) then         !stomatal conductance calculated from Ball et al. (1987)
+               else if ( stomatal_model == 1 ) then         !stomatal conductance calculated from Ball et al. (1987)
+                  if (use_agross == 1) then
```
Comment (rgknox) [inline biogeophys/FatesPlantRespPhotosynthMod.F90:1167]: I can't find where use_agross is defined.  
Should we modify this logic anyway to just use agross if LWP is provided?

File: biogeophys/FatesPlantRespPhotosynthMod.F90 (hunk 8 above)
Comment (rgknox) [reply biogeophys/FatesPlantRespPhotosynthMod.F90:1279]: Could we also change 18.0 and 8.314 to named parameter constants defined at the head of this routine?

Comment (ckoven) [reply biogeophys/FatesPlantRespPhotosynthMod.F90:1279]: or better yet, https://github.com/NGEET/fates/blob/master/main/FatesConstantsMod.F90#L184

File: biogeophys/FatesPlantRespPhotosynthMod.F90 [hunk 15]
```diff
@@ -1214,17 +1248,70 @@ subroutine LeafLayerPhotosynthesis(f_sun_lsl,         &  ! in
...
-           rstoma_out = 1._r8/gstoma
-	   
+            ! This is the stomatal resistance of the leaf layer of a given cohort
+
+            ! Adjusting gs (compute a virtual gs) that will be passed to host model
+            ! compute specific humidity from vapor pressure
+            ! q = 0.622*e/(can_press - 0.378*e) 
+            ! source https://cran.r-project.org/web/packages/humidity/vignettes/humidity-measures.html
+
+           if (hlm_use_planthydro.eq.itrue) then
```
Comment (rgknox) [inline biogeophys/FatesPlantRespPhotosynthMod.F90:1265]: Might be a little cleaner if we change this to:
> if (hlm_use_planthydro.eq.itrue .and. EDPftvarcon_inst%hydr_k_lwp(ft)>nearzero  ) then

then remove the inner logicals k_lwp

I'd also recommend bounding all of the logic here in its own subroutine or function call

File: biogeophys/FatesPlantRespPhotosynthMod.F90 (hunk 14 above)
Comment (rgknox) [reply biogeophys/FatesPlantRespPhotosynthMod.F90:1167]: Notes taken with @JunyanDing during zoom chat:
```
logical, parameter :: use_agross = .false.

if (use_agross .or. (  (hlm_use_planthydro.eq.itrue).and.     (EDPftvarcon_inst%hydr_k_lwp(ft)>0.))) then
```

File: biogeophys/FatesPlantRespPhotosynthMod.F90 [hunk 16]
```diff
@@ -1214,17 +1267,70 @@ subroutine LeafLayerPhotosynthesis(f_sun_lsl,         &  ! in
...
+             qsat_adj = qsat*LWP_star
+             
+             if (k_lwp == 0 ) then 
+               rstoma_out = 1._r8/gstoma
+             else 
+               if ((qsat_adj - qs) <= 0) then
+          
+                 ! if inner leaf vapor pressure is less then or equal to that at leaf surface
+                 ! then set stomata resistance to be very large to stop the transpiration or back flow of vapor
+                  rstoma_out = rsmax0*100
```
[the remark this answers is not in the document]
Comment (rgknox) [reply biogeophys/FatesPlantRespPhotosynthMod.F90]: discussed with @JunyanDing and we are reverted to rsmax0

**Comment by rgknox** [comment]
Tests look good, only known fails, b4b with base.
```
rgknox/tests_1102-100244ch> ./cs.status | grep FAIL
    FAIL SMS_Lm13.1x1_brazil.I2000Clm50FatesCruRsGs.cheyenne_gnu.clm-FatesColdDef TPUTCOMP Error: TPUTCOMP: Computation time increase > 25% from baseline
  ERS_Ld30.f45_f45_mg37.I2000Clm50FatesCruRsGs.cheyenne_intel.clm-FatesColdDefReducedComplexSatPhen (Overall: FAIL) details:
    FAIL ERS_Ld30.f45_f45_mg37.I2000Clm50FatesCruRsGs.cheyenne_intel.clm-FatesColdDefReducedComplexSatPhen RUN time=60
  ERS_Lm12.1x1_brazil.I2000Clm50FatesCruRsGs.cheyenne_intel.clm-FatesFireLightningPopDens (Overall: FAIL) details:
    FAIL ERS_Lm12.1x1_brazil.I2000Clm50FatesCruRsGs.cheyenne_intel.clm-FatesFireLightningPopDens COMPARE_base_rest
    FAIL ERS_Lm12.1x1_brazil.I2000Clm50FatesCruRsGs.cheyenne_intel.clm-FatesFireLightningPopDens TPUTCOMP Error: TPUTCOMP: Computation time increase > 25% from baseline
```
